File: backend/app/api/records.py
"""Direct editing of the boards, without going through the agent.

Everything on screen was previously read-only unless you talked to JARVIS about
it. That is fine for "add a task to call the dentist" and absurd for fixing a
typo, so this exposes the same four stores the tools operate on:

    tasks  ·  schedules  ·  ideas  ·  memories

Deliberately the *same* `crud` functions the tool handlers call, not a parallel
implementation. The rules about what an edit means — completing a task removes
it from the board, a delete leaves a tombstone so sync cannot resurrect it,
priorities coerce rather than reject — have to hold identically whether the
change came from a person or the model. Two implementations would drift, and
the drift would only show up days later as a row that came back from the dead.

Only meaningful on desktop. When the client owns the data (mobile) the frontend
writes to IndexedDB directly and never calls these; see
`frontend/src/lib/localMutations.ts`, which mirrors the same semantics for the
same reason.
"""

from __future__ import annotations

import logging

from fastapi import APIRouter, HTTPException

from app.api.schemas import (
    IdeaCreate,
    IdeaOut,
    IdeaUpdate,
    MemoryCreate,
    MemoryOut,
    MemoryUpdate,
    NotePageSave,
    ScheduleEventCreate,
    ScheduleEventOut,
    ScheduleEventUpdate,
    TaskCreate,
    TaskOut,
    TaskUpdate,
)
from app.db import crud
from app.db.database import db

logger = logging.getLogger("jarvis.api.records")

router = APIRouter(prefix="/api", tags=["records"])


def _present(payload, *, drop: set[str] = frozenset()) -> dict:
    """Only the fields the caller actually sent.

    `exclude_unset` is what makes a partial update partial: without it, every
    unmentioned field arrives as None and a rename would blank the rest of the
    record.
    """
    return {
        key: value
        for key, value in payload.model_dump(exclude_unset=True).items()
        if key not in drop
    }



#: Path identifier -> table, for the uid resolver below.
_TABLES = {"task": "tasks", "event": "schedules", "idea": "ideas", "memory": "memories"}


async def _rowid(kind: str, ident: str) -> int:
    """Resolve a path identifier that may be a uid or a rowid.

    A rowid is not a stable name for a record here, and treating it as one
    produced a bug that looked impossible: deleting a task you were looking at
    answered "No task with id N."

    The reason is the client-owned-data working copy. Seeding it runs
    ``DELETE FROM tasks`` followed by an INSERT of the synced columns -- and
    ``SYNC_COLUMNS`` does not include ``id``, so SQLite assigns brand new
    rowids every single time. Seeding happens on launch and after every hand
    edit; one device log showed twenty reseeds in two days. Any screen holding
    an id from before a reseed is addressing a number that has since moved to a
    different row, or to no row at all.

    ``uid`` is the identity that survives, so the client now sends that and this
    turns it back into whatever rowid the working copy currently uses. Plain
    integers still work unchanged, which keeps the desktop and the existing
    tests on exactly the path they were on.
    """
    text = str(ident).strip()
    table = _TABLES[kind]

    if text.isdigit():
        return int(text)

    row = await db.fetch_one(f"SELECT id FROM {table} WHERE uid = ?", (text,))
    if row is None:
        raise HTTPException(404, f"No {kind} with uid {text}.")
    return int(row["id"])


# --- tasks -----------------------------------------------------------------

@router.post("/tasks", response_model=TaskOut, status_code=201)
async def create_task(payload: TaskCreate) -> TaskOut:
    task = await crud.create_task(
        title=payload.title,
        due_date=payload.due_date,
        priority=payload.priority,
        category=payload.category,
    )
    return TaskOut(**task)


@router.patch("/tasks/{task_id}", response_model=TaskOut)
async def update_task(task_id: str, payload: TaskUpdate) -> TaskOut:
    fields = _present(payload, drop={"clear_due_date"})
    if payload.clear_due_date:
        fields["due_date"] = None

    change = await crud.update_task(await _rowid("task", task_id), **fields)
    if change.task is None:
        raise HTTPException(404, f"No task with id {task_id}.")
    if change.cleared:
        # Completing removes the task, so there is no row left to return. Say
        # so plainly rather than handing back a record that no longer exists.
        raise HTTPException(
            409,
            f"'{change.task['title']}' was completed and cleared off the board.",
        )
    return TaskOut(**change.task)


@router.delete("/tasks/{task_id}", status_code=204)
async def delete_task(task_id: str) -> None:
    if await crud.delete_task(await _rowid("task", task_id)) is None:
        raise HTTPException(404, f"No task with id {task_id}.")


# --- schedule --------------------------------------------------------------

@router.post("/schedule", response_model=ScheduleEventOut, status_code=201)
async def create_event(payload: ScheduleEventCreate) -> ScheduleEventOut:
    event = await crud.create_schedule_event(**payload.model_dump())
    return ScheduleEventOut(**event)


@router.patch("/schedule/{event_id}", response_model=ScheduleEventOut)
async def update_event(event_id: str, payload: ScheduleEventUpdate) -> ScheduleEventOut:
    event = await crud.update_schedule_event(await _rowid("event", event_id), **_present(payload))
    if event is None:
        raise HTTPException(404, f"No schedule event with id {event_id}.")
    return ScheduleEventOut(**event)


@router.delete("/schedule/{event_id}", status_code=204)
async def delete_event(event_id: str) -> None:
    if await crud.delete_schedule_event(await _rowid("event", event_id)) is None:
        raise HTTPException(404, f"No schedule event with id {event_id}.")


# --- ideas -----------------------------------------------------------------

@router.post("/ideas", response_model=IdeaOut, status_code=201)
async def create_idea(payload: IdeaCreate) -> IdeaOut:
    idea = await crud.create_idea(
        title=payload.title,
        description=payload.description,
        tags=payload.tags,
        status=payload.status,
        page_uid=payload.page_uid,
    )
    return IdeaOut(**idea)


@router.patch("/ideas/{idea_id}", response_model=IdeaOut)
async def update_idea(idea_id: str, payload: IdeaUpdate) -> IdeaOut:
    idea = await crud.update_idea(await _rowid("idea", idea_id), **_present(payload))
    if idea is None:
        raise HTTPException(404, f"No idea with id {idea_id}.")
    return IdeaOut(**idea)


@router.delete("/ideas/{idea_id}", status_code=204)
async def delete_idea(idea_id: str) -> None:
    if await crud.delete_idea(await _rowid("idea", idea_id)) is None:
        raise HTTPException(404, f"No idea with id {idea_id}.")


# --- memories --------------------------------------------------------------

@router.post("/memories", response_model=MemoryOut, status_code=201)
async def create_memory(payload: MemoryCreate) -> MemoryOut:
    # upsert, not insert: memories are keyed by concept, and saving the same
    # concept twice should correct it rather than leave two contradictory rows
    # for the model to pick between.
    memory = await crud.upsert_memory(
        key_concept=payload.key_concept,
        content=payload.content,
        category=payload.category,
        expires_at=payload.expires_at,
        memory_type=payload.memory_type,
        memory_status=payload.memory_status,
        confidence=payload.confidence,
        importance=payload.importance,
        pinned=payload.pinned,
        tags=payload.tags,
    )
    return MemoryOut(**memory)


@router.patch("/memories/{memory_id}", response_model=MemoryOut)
async def update_memory(memory_id: str, payload: MemoryUpdate) -> MemoryOut:
    memory = await crud.update_memory(await _rowid("memory", memory_id), **_present(payload))
    if memory is None:
        raise HTTPException(404, f"No memory with id {memory_id}.")
    return MemoryOut(**memory)


@router.delete("/memories/{memory_id}", status_code=204)
async def delete_memory(memory_id: str) -> None:
    if await crud.delete_memory(await _rowid("memory", memory_id)) is None:
        raise HTTPException(404, f"No memory with id {memory_id}.")


@router.post("/note-pages")
async def save_note_page(payload: NotePageSave) -> dict:
    if not payload.title.strip():
        raise HTTPException(422, "Give this page a name.")
    return await crud.save_note_page(payload.uid, payload.title, payload.kind)


@router.delete("/note-pages/{page_uid}", status_code=204)
async def delete_note_page(page_uid: str) -> None:
    if not await crud.delete_note_page(page_uid):
        raise HTTPException(404, "Only custom Notes pages can be deleted.")

File: frontend/src/lib/records.ts
/**
 * Creating, editing and deleting records by hand.
 *
 * Everything on the boards used to be read-only unless you asked JARVIS to
 * change it, which is fine for "add a task to call the dentist" and absurd for
 * fixing a typo. This is the layer that makes them editable.
 *
 * It exists because there are two authoritative stores and the components must
 * not know which one they are talking to. On the desktop the backend's SQLite
 * is the truth and these go over HTTP; on mobile IndexedDB is the truth and
 * they go straight to the local store and queue for sync. Every function here
 * takes the same arguments either way, so a component never branches on it —
 * exactly the split `handleToggleTask` already makes, generalised so it does
 * not have to be re-made for every field.
 *
 * The semantics on both sides are the ones in `backend/app/db/crud.py`, because
 * the same record is editable from either device and from the agent. Completing
 * a task removes it; a delete leaves a tombstone. Two implementations that
 * disagreed would show up days later as a row returning from the dead.
 */

import { apiDelete, apiPatch, apiPost } from "@/lib/api";
import {
  deleteRowLocally,
  updateRowLocally,
  type RecordRef,
} from "@/lib/localMutations";
import { type SyncedTable, listRows, putRows } from "@/lib/localdb";
import { numericId } from "@/lib/localDashboard";
import { decodeMemoryContent, encodeMemoryContent, reviseMemoryContent } from "@/lib/memory";
import { nowIso } from "@/lib/syncClient";
import type { Idea, Memory, NotePage, Reminder, ScheduleEvent, Task } from "@/types";

/** A fresh uid for a locally-created row, matching the backend's format. */
function newUid(): string {
  const cryptoRef = globalThis.crypto;
  if (cryptoRef?.randomUUID) return cryptoRef.randomUUID();
  // Older WebViews: good enough for a local identifier, and the uid only has
  // to be unique, never unguessable.
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === "x" ? r : (r & 0x3) | 0x8).toString(16);
  });
}

async function createLocally(
  table: SyncedTable,
  fields: Record<string, unknown>,
): Promise<void> {
  const stamp = nowIso();
  await putRows(table, [
    { uid: newUid(), created_at: stamp, updated_at: stamp, ...fields } as never,
  ]);
}

export interface RecordsMode {
  /** True when IndexedDB is authoritative (the mobile build). */
  local: boolean;
}

/* -------------------------------------------------------------------------- */
/* Tasks                                                                       */
/* -------------------------------------------------------------------------- */

export interface TaskDraft {
  title: string;
  due_date?: string | null;
  priority?: Task["priority"];
  category?: string | null;
}

export async function createTask({ local }: RecordsMode, draft: TaskDraft): Promise<void> {
  if (local) {
    await createLocally("tasks", {
      title: draft.title.trim(),
      category: (draft.category || "GENERAL").toUpperCase(),
      priority: draft.priority ?? "MEDIUM",
      status: "PENDING",
      due_date: draft.due_date ?? null,
    });
    return;
  }
  await apiPost("/api/tasks", draft);
}

export interface TaskEdit extends Partial<TaskDraft> {
  status?: Task["status"];
  clear_due_date?: boolean;
}

export async function updateTask(
  { local }: RecordsMode,
  id: RecordRef,
  changes: TaskEdit,
): Promise<void> {
  if (local) {
    const { clear_due_date, ...rest } = changes;
    await updateRowLocally("tasks", id, {
      ...rest,
      ...(clear_due_date ? { due_date: null } : {}),
    });
    return;
  }
  await apiPatch(`/api/tasks/${id}`, changes);
}

export async function deleteTask({ local }: RecordsMode, id: RecordRef): Promise<void> {
  if (local) {
    await deleteRowLocally("tasks", id);
    return;
  }
  await apiDelete(`/api/tasks/${id}`);
}

/* -------------------------------------------------------------------------- */
/* Schedule                                                                    */
/* -------------------------------------------------------------------------- */

export interface EventDraft {
  event_name: string;
  kind?: ScheduleEvent["kind"];
  day_of_week?: number | null;
  start_time?: string | null;
  end_time?: string | null;
  time_start?: string | null;
  time_end?: string | null;
  location?: string | null;
  notes?: string | null;
}

export async function createEvent({ local }: RecordsMode, draft: EventDraft): Promise<void> {
  if (local) {
    await createLocally("schedules", {
      event_name: draft.event_name.trim(),
      kind: draft.kind ?? "SESSION",
      day_of_week: draft.day_of_week ?? null,
      start_time: draft.start_time ?? null,
      end_time: draft.end_time ?? null,
      time_start: draft.time_start ?? null,
      time_end: draft.time_end ?? null,
      location: draft.location ?? null,
      notes: draft.notes ?? null,
    });
    return;
  }
  await apiPost("/api/schedule", draft);
}

export async function updateEvent(
  { local }: RecordsMode,
  id: RecordRef,
  changes: Partial<EventDraft>,
): Promise<void> {
  if (local) {
    await updateRowLocally("schedules", id, changes);
    return;
  }
  await apiPatch(`/api/schedule/${id}`, changes);
}

export async function deleteEvent({ local }: RecordsMode, id: RecordRef): Promise<void> {
  if (local) {
    await deleteRowLocally("schedules", id);
    return;
  }
  await apiDelete(`/api/schedule/${id}`);
}

/* -------------------------------------------------------------------------- */
/* Ideas                                                                       */
/* -------------------------------------------------------------------------- */

export interface IdeaDraft {
  page_uid?: string | null;
  title: string;
  description?: string;
  tags?: string;
  status?: Idea["status"];
}

export async function createIdea({ local }: RecordsMode, draft: IdeaDraft): Promise<void> {
  if (local) {
    await createLocally("ideas", {
      title: draft.title.trim(),
      description: draft.description ?? "",
      tags: draft.tags ?? "",
      status: draft.status ?? "DRAFT",
      page_uid: draft.page_uid ?? null,
    });
    return;
  }
  await apiPost("/api/ideas", draft);
}

export async function updateIdea(
  { local }: RecordsMode,
  id: RecordRef,
  changes: Partial<IdeaDraft>,
): Promise<void> {
  if (local) {
    await updateRowLocally("ideas", id, changes);
    return;
  }
  await apiPatch(`/api/ideas/${id}`, changes);
}

export async function deleteIdea({ local }: RecordsMode, id: RecordRef): Promise<void> {
  if (local) {
    await deleteRowLocally("ideas", id);
    return;
  }
  await apiDelete(`/api/ideas/${id}`);
}

/* -------------------------------------------------------------------------- */
/* Memories                                                                    */
/* -------------------------------------------------------------------------- */

export interface MemoryDraft {
  expires_at?: string | null;
  key_concept: string;
  content: string;
  category?: Memory["category"];
  memory_type?: Memory["memory_type"];
  memory_status?: Memory["memory_status"];
  confidence?: number;
  importance?: number;
  pinned?: boolean;
  tags?: string[];
}

export async function createMemory({ local }: RecordsMode, draft: MemoryDraft): Promise<void> {
  if (local) {
    const stamp = nowIso();
    await createLocally("memories", {
      key_concept: draft.key_concept.trim(),
      content: encodeMemoryContent(draft.content, {
        type: draft.memory_type,
        status: draft.memory_status,
        confidence: draft.confidence,
        importance: draft.importance,
        pinned: draft.pinned,
        tags: draft.tags,
        source_kind: "manual_ui",
        valid_from: stamp,
      }),
      category: draft.category ?? "LONG_TERM",
      expires_at: draft.expires_at ?? null,
    });
    return;
  }
  await apiPost("/api/memories", draft);
}

export async function updateMemory(
  { local }: RecordsMode,
  id: RecordRef,
  changes: Partial<MemoryDraft>,
): Promise<void> {
  if (local) {
    const rows = await listRows("memories");
    const existing = rows.find(row => typeof id === "string" ? row.uid === id : numericId(row.uid) === id);
    if (!existing) return;
    const decoded = decodeMemoryContent(existing.content, {
      category: existing.category,
      expiresAt: existing.expires_at,
      createdAt: existing.created_at,
    });
    const body = changes.content ?? decoded.body;
    const content = reviseMemoryContent(
      existing.content,
      body,
      {
        type: changes.memory_type,
        status: changes.memory_status,
        confidence: changes.confidence,
        importance: changes.importance,
        pinned: changes.pinned,
        tags: changes.tags,
        source_kind: "manual_ui",
      },
      {
        category: changes.category ?? existing.category,
        expiresAt: changes.expires_at ?? existing.expires_at,
        createdAt: existing.created_at,
        updatedAt: existing.updated_at,
      },
    );
    const storedChanges: Record<string, unknown> = {};
    if (changes.key_concept !== undefined) storedChanges.key_concept = changes.key_concept;
    if (changes.category !== undefined) storedChanges.category = changes.category;
    if (changes.expires_at !== undefined) storedChanges.expires_at = changes.expires_at;
    storedChanges.content = content;
    await updateRowLocally("memories", id, storedChanges);
    return;
  }
  await apiPatch(`/api/memories/${id}`, changes);
}

export async function deleteMemory({ local }: RecordsMode, id: RecordRef): Promise<void> {
  if (local) {
    await deleteRowLocally("memories", id);
    return;
  }
  await apiDelete(`/api/memories/${id}`);
}

export async function saveNotePage({local}: RecordsMode, page: Pick<NotePage, "title" | "kind"> & Partial<NotePage>): Promise<NotePage> {
  const stamp = nowIso();
  const row = {...page, title: page.title.trim(), uid: page.uid || newUid(), created_at: page.created_at || stamp, updated_at: stamp} as NotePage;
  if (!row.title) throw new Error("Give this page a name.");
  if (local) { await putRows("note_pages", [row as unknown as import("@/lib/localdb").SyncRow]); return row; }
  return apiPost<NotePage>("/api/note-pages", row);
}

export async function deleteNotePage({local}: RecordsMode, page: NotePage): Promise<void> {
  if (page.kind !== "CUSTOM") throw new Error("Default pages cannot be deleted.");
  if (local) {
    const ideas = await import("@/lib/localdb").then(module => module.listRows("ideas"));
    const moved = ideas.filter(row => row.page_uid === page.uid).map(row => ({...row, page_uid: null, updated_at: nowIso()}));
    if (moved.length) await putRows("ideas", moved);
    await deleteRowLocally("note_pages", page.uid);
    return;
  }
  await apiDelete(`/api/note-pages/${page.uid}`);
}

/* -------------------------------------------------------------------------- */
/* Reminders                                                                   */
/*                                                                              */
/* No `RecordsMode` here, on purpose. Unlike the five stores above, reminders  */
/* are never mirrored into IndexedDB or Supabase — each device fires its own  */
/* alarms off its own local backend (see `fetchReminders` in `lib/api.ts`),   */
/* so there is only ever one way to reach them, on desktop or mobile alike.   */
/* -------------------------------------------------------------------------- */

export interface ReminderDraft {
  text: string;
  /** Absolute datetime, ISO 8601 local. */
  due_at: string;
}

export async function createReminder(draft: ReminderDraft): Promise<Reminder> {
  return apiPost<Reminder>("/api/reminders", draft);
}

export async function updateReminder(
  id: number,
  changes: Partial<ReminderDraft>,
): Promise<Reminder> {
  return apiPatch<Reminder>(`/api/reminders/${id}`, changes);
}

export async function deleteReminder(id: number): Promise<void> {
  await apiDelete(`/api/reminders/${id}`);
}
